### PR TITLE
FOXSI New Startup

### PR DESCRIPTION
**Pull request by pymilo**
A new way to startup FOXSI. The README still need to be changed.

### CODE REVIEW

**Comment by jtvievering** [comment]
Milo and I have been working this week to make the FOXSI analysis setup a little more friendly and flexible for users. What this will entail is including a few lines in your personal IDL startup script and calling the setup procedure in a different way, and then you should be able to run the FOXSI analysis from anywhere you'd like, rather than having to be in the foxsi-science folder. This will have a minimal effect on users, but we just wanted to check if there are any objections/suggestions! Please let us know by the end of the day tomorrow (Wednesday) if you have any comments. After that, we will merge this pull request and make the necessary changes to other files to implement this plan.

File: README.md [hunk 1]
```diff
@@ -1,6 +1,28 @@
 foxsi-science
 =============
 
+INTALL:
```
Comment (LinErinG) [inline README.md]: Typo: INSTALL

File: foxsi.pro [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+	endif else begin
+		print, 'FOXSI did not fly that year.'
+		print, 'Please use one of the next years:'
+		print, '       - 2012'
+		print, '       - 2014'
+		print, 'Example: foxsi,2014'
+	endelse
+endelse
+stop
```
Comment (LinErinG) [inline foxsi.pro]: I don't think we want to stop within the routine.  (Otherwise, every time we later "retall" to return to the main level we'll lose our parameters.)  We should find a way to make these parameters available to the main level.

File: foxsi-setup-script-2012.pro [hunk 3]
```diff
@@ -1,83 +0,0 @@
...
-; Save the values set here for use by some of the routines.
-; Saving this file each time ensures that updating a value in this setup script
-; propagates through to all routines.
-save, tlaunch, t_launch, t1_start, t1_end, t2_start, t2_end, t3_start, t3_end, $
-	t4_start, t4_end, t5_start, t5_end, t6_start, t6_end, date, t0, $	
-	offset_xy, cen1, cen2, cen3, cen4, cen5, cen6, $
-	shift0, shift1, shift2, shift3, shift4, shift5, shift6, $
-	rot0, rot1, rot2, rot3, rot4, rot5, rot6, $
-	detnum0, detnum1, detnum2, detnum3, detnum4, detnum5, detnum6, $
-	file = 'data_2012/flight2012-parameters.sav'
```
Comment (LinErinG) [inline foxsi-setup-script-2012.pro]: I agree with removing this file completely (as well as the 2014 file).  We can't keep them for backwards compatibility because it would be tedious to try to keep them up-to-date with any changed parameters.

File: util/foxsi-setup-script-2014.pro [hunk 4]
```diff
@@ -126,4 +114,4 @@ save, tlaunch, t_launch, t1_pos0_start, t1_pos0_end, t1_pos1_start, $
 	shift0, shift1, shift2, shift3, shift4, shift5, shift6, rot0, rot1, rot2, rot3, $
 	rot4, rot5, rot6, $
 	detnum0, detnum1, detnum2, detnum3, detnum4, detnum5, detnum6, $
-	file = 'data_2014/flight2014-parameters.sav'
+	file = FOXSIDB+'/data_2014/flight2014-parameters.sav'
```
Comment (LinErinG) [inline util/foxsi-setup-script-2014.pro]: Is there a reason why this file is getting kept but the old 2012 setup file is getting scrapped?

File: README.md [hunk 5]
```diff
@@ -1,6 +1,28 @@
...
+    
+    Include this five lines in your personal IDL_STARTUP file
+
+    COMMON FOXSI,FOXSIPKG,FOXSIDB
+    FOXSIPKG='path/to/your/foxsi-science/directory'
+    FOXSIDB='path/to/your/foxsi-data/directory/foxsidb'
+    add_path,FOXSIPKG
+    add_path,FOXSIDB
```
Comment (LinErinG) [inline README.md]: We should add some guidance as to which files should be put within the foxsidb folder.  

Side note, just in case it's useful -- the add_path procedure has a keyword /expand that can be used to add all subdirectories in addition to the directory you specify.

File: foxsi.pro [hunk 6]
```diff
@@ -0,0 +1,42 @@
+; June 2016
+; STARTUP FOXSI
+; Description: 
+;
+;
+
+PRO FOXSI, YEAR
+
+COMMON FOXSI
```
Comment (ehsteve) [inline foxsi.pro]: FYI, I've gotten some feedback that people really don't like our use of common. Not sure if there is any way around it though.

File: util/foxsi-setup-script-2014.pro (hunk 4 above)
Comment (ehsteve) [inline util/foxsi-setup-script-2014.pro]: I would recommend we not use .sav files at all. Not very friendly on github or anybody else.

Comment (ehsteve) [inline util/foxsi-setup-script-2014.pro]: Can this just be put in foxsi.pro?

File: foxsi.pro (hunk 6 above)
Comment (pymilo) [inline foxsi.pro]: I figured that it may be a way around it that involves using '.sav' files, but I already read that you don't like that option either. Do you have any advice?

File: README.md [hunk 7]
```diff
@@ -1,16 +1,55 @@
...
+
+    Include this five lines in your personal IDL_STARTUP file
+
+    COMMON FOXSI,FOXSIPKG,FOXSIDB
+    FOXSIPKG='path/to/your/foxsi-science'
+    FOXSIDB='path/to/your/foxsi-data/directory/foxsidb'
+    add_path,FOXSIPKG,/expand
+    add_path,FOXSIDB
+    restore,FOXSIDB+'/data_2014/flight2014-parameters.sav'
+
```
Comment (jtvievering) [inline README.md]: Using the alternate method we discussed, lines 10-15 would be replaced with:

setenv, 'FOXSIPKG=/path/to/your/foxsi-science'
setenv, 'FOXSIDB=/path/to/your/foxsi-data/directory/foxsidb'
add_path, '$FOXSIPKG', /expand
add_path, '$FOXSIDB'
restore, '$FOXSIDB/data_2014/flight2014-parameters.sav'

File: foxsi.pro (hunk 6 above)
Comment (ehsteve) [inline foxsi.pro]: As discussed during the telecon, I think the path forward we agreed on was to use txt files for as much as is appropriate.

File: README.md [hunk 8]
```diff
@@ -1,40 +1,77 @@
 foxsi-science
 =============
 
+INSTALL:
+
+    You need to have IDL and SSW previously installed
+
+    Include this five lines in your personal IDL_STARTUP file
```
Comment (LinErinG) [inline README.md]: You should also instruct the user to change these lines to reflect their own directory structure, and tell them what to change.  For example, ~/Desktop/foxsi-science should be changed to whatever path is needed.  With the next line, is WORK_SSWIDL/Default/foxsidb something the user needs to change or is WORK_SSWIDL an environment variable that SSWIDL already knows at this point?  Give the user very specific instructions on how to change these lines.

File: foxsi.pro [hunk 9]
```diff
@@ -0,0 +1,57 @@
+; June 2016
```
Comment (LinErinG) [inline foxsi.pro]: Is this file still necessary?

File: foxsi.pro [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+			; load the Level 2 data.
+			@param2014
+			restore, '$FOXSIDB'+'/data_2014/foxsi_level2_data.sav', /v
+			DATA_LVL2_D0=DATA_LVL2_D0
+			DATA_LVL2_D1=DATA_LVL2_D1
+			DATA_LVL2_D2=DATA_LVL2_D2
+			DATA_LVL2_D3=DATA_LVL2_D3
+			DATA_LVL2_D4=DATA_LVL2_D4
+			DATA_LVL2_D5=DATA_LVL2_D5
+			DATA_LVL2_D6=DATA_LVL2_D6
```
Comment (LinErinG) [inline foxsi.pro]: I don't think you need these lines, since they don't change the variables.

**Comment by LinErinG** [comment]
@pymilo @jtvievering @ehsteve @aringlis 

I think this is almost done!  I have one more request, though:

Please rename the param2012 and param2014 files to include a date, for example, "param2014_20160620" or something similar.  (Using today's date would be fine for the initial version.)  That will be our version control for our configuration parameters.  Due to the philosophy you've implemented, when we want to change a parameter for the future, we only have to add a file with a new date, edit one line in foxsi.pro, and update GitHub.

Since it looks like this is converging, please write up a short email that describes what tactic you chose for the configuration parameters and why that was chosen over the alternatives.  Send that out to the team to get feedback, and encourage everyone to look at (and try) this code.  May be best to set a deadline on that so we can wrap this up relatively quickly.

File: img/draw_fov.pro [hunk 11]
```diff
@@ -20,7 +20,8 @@ PRO		DRAW_FOV, DET=DET, ALL=ALL, GOOD=GOOD, TARGET=TARGET, XYCOR=XYCOR, YEAR=YEA
 ;
 
 
-	default, year, 2014
+default, year, 2014
+COMMON PARAM
```
Comment (ehsteve) [inline img/draw_fov.pro]: this seems like a dangerously vanilla name for a common block. Could it be changed to FOXSI_PARAM or something?

File: parameters/param2012_20160620.pro [hunk 12]
```diff
@@ -0,0 +1,45 @@
+t_launch = 64500.
```
Comment (ehsteve) [inline parameters/param2012_20160620.pro]: I would remove the date from the filename as version control will already know when it was changed. This kind of "version control" does not play nice with git as you'd lose the history of the file every time you updated it. You would need a new file with a new date on it. If you really want to keep some sort of revision date in the file then maintain as a comment inside the file and not in the filename. It might also confuse a user since there are two dates in that filename, one refers to FOXSI and the other to the last revision date (which they don't really need to know).

File: parameters/param2014_20160620.pro [hunk 13]
```diff
@@ -0,0 +1,69 @@
+t_launch = 69060
```
Comment (ehsteve) [inline parameters/param2014_20160620.pro]: same here!

**Comment by ehsteve** [comment]
I read the email about the use of COMMON blocks and I am convinced that it is the right way to go for this particular application. As long as my two comments are dealt with I am 👍.

**Comment by LinErinG** [comment]
Thanks, @ehsteve, for looking at this.  Since the filename choice was at my request, let me address that.  What I'd like is for there to be extremely explicit version control on the configuration parameter file so that the user knows exactly what instrument config they're using.  Changing a parameter in the configuration file is a more "serious" alteration than making most other code changes.  Let's say, for example, that someone used the analysis code last November, and uses it again next year.  If the config file was changed in that time, they might get different results because, for example, we fine-tuned the target pointing knowledge.  Such a discrepancy could be easily trouble-shot by asking the question "what configuration file are you using?"  If the user has all the past configurations locally, they can easily drop in another configuration file from another point in history to compare the effects of changes.  This was partly inspired by the way NuSTAR does their calibration databases -- with a "calibration publication" date hard-coded in the index file name so that it's easy to tell what version is being used.  If we do this a ton, it would get out of control, but I hope that we're not going to be changing these files too often.  What are your thoughts on this?  (P.S. I'm also thinking it would be good to do this with the data files and some calibration files, too.)

**Comment by ehsteve** [comment]
@LinErinG, I see what you are trying to accomplish. That would make it easier than requiring people to mess with git to grab old versions of the calibration. 👍

**Comment by LinErinG** [comment]
Sounds like there's general agreement, so I'll merge this!  Richard has some further suggestions that I'll discuss at the meeting today, but those could be implemented as a future edit.

**Comment by jtvievering** [comment]
Did this get merged before it was closed? I'm not seeing the changes in foxsi-science.
